src/query/service/src/interpreters/interpreter_select_v2.rs

```diff
@@ -185,22 +200,43 @@ impl Interpreter for SelectInterpreterV2 {
             let key = gen_result_cache_key(self.formatted_ast.as_ref().unwrap());
             // 1. Try to get result from cache.
             let kv_store = UserApiProvider::instance().get_meta_store_client();
-            let cache_reader = ResultCacheReader::create(
-                self.ctx.clone(),
-                &key,
-                kv_store.clone(),
-                self.ctx
-                    .get_settings()
-                    .get_query_result_cache_allow_inconsistent()?,
-            );
+
+            // Execute `select * from result_scan(last_query_id)` multiple times
+            // should return same result. Please consider the following scenarios:
+            // 1) select * from t1;
+            // 2) select * from result_scan(last_query_id()); --> returns result same as line 1
+            // 3) insert into t1 values(2);
+            // 4) select * from t1; --> result changed since we insert new data.
+            // 5) select * from result_scan(last_query_id()); --> result same as line 2 cause cache
+            // If we read cache for 5, we will see it returns same result as 1 and 2 cause the
+            // generated result_cache_key are same for this statement, so here we fetch the previous
+            // meta_key through related query_id and using this meta_key to read cache.
+            let meta_key = if let Some(t) = self.result_scan_table() {
+                let arg_query_id = parse_result_scan_args(&t.table_args().unwrap())?;
+                self.ctx.get_result_cache_key(&arg_query_id)
```

Comment:
After creating `ResultScanTable`, we can use the original pipeline (don't need to create cache reader again).